### PR TITLE
Add basics of functional sims

### PR DESCRIPTION
**Pull request by cesarus777**
Add simulator base class for different interpreter and switched sim as example. Also add ISA module and instruction classes.

### CODE REVIEW

File: FunctionalSim/FunctionalSim.hpp [hunk 1]
```diff
@@ -0,0 +1,383 @@
...
+
+  void SPIncrement() {
+    SP += cellsPerReg;
+    if (SP > (ram->getProgramEnd() + RAM::stackOffset)) {
+      throw shisa::StackOverflow{};
+    }
+  }
+
+public:
+  Cpu() {
```
Comment (GreenSeal) [inline FunctionalSim/FunctionalSim.hpp:180]: Why don't you write destructor here?

File: FunctionalSim/FunctionalSim.hpp [hunk 2]
```diff
@@ -0,0 +1,383 @@
...
+  shisa::ISAModule *ISAModule;
+
+  using cpu_t = Cpu<addr_t, memCell_t, reg_t, nRegs, MemStorage>;
+  cpu_t *state;
+
+protected:
+  auto getState() -> cpu_t * { return state; }
+
+public:
+  SimBase(const shisa::ISAModule &m) {
```
Comment (GreenSeal) [inline FunctionalSim/FunctionalSim.hpp:270]: And here?

File: FunctionalSim/FunctionalSim.hpp [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+    }
+
+    return addr;
+  }
+
+  bool   binaryLoaded = false;
+  addr_t programEnd   = 0;
+
+public:
+  MemModel() = default;
```
Comment (GreenSeal) [inline FunctionalSim/FunctionalSim.hpp]: Why do wee need this line?

File: FunctionalSim/FunctionalSim.hpp (hunk 1 above)
Comment (cesarus777) [reply FunctionalSim/FunctionalSim.hpp:180]: Forgot)

File: FunctionalSim/FunctionalSim.hpp (hunk 2 above)
Comment (cesarus777) [reply FunctionalSim/FunctionalSim.hpp:270]: And here forgot)

File: FunctionalSim/FunctionalSim.hpp (hunk 3 above)
Comment (cesarus777) [reply FunctionalSim/FunctionalSim.hpp]: It's a kind of legacy. I wanted to add memory initialization in this constructor, but then removed binary loading to a separate function and marked the constructor explicitly as default. Agree it should be removed.

**Pull request merged**